### PR TITLE
DS integration in Webhooks Listview 

### PR DESCRIPTION
**Pull request by ronronscelestes**
## What does it do?

Integration of DS components in Settings/Webhooks Listview 💅
Actions still work using actions logic that was already there - refacto after this PR
Old List component left in comments for future implementation of actions (as PopUpWarning)

## Demo
Run this branch
Capture d’écran 2021-08-13 à 17 10 35
Capture d’écran 2021-08-13 à 17 12 19

### CODE REVIEW

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 1]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
+              <Thead>
+                <Tr>
+                  <Th>
+                    <BaseCheckbox
+                      aria-label="Select all entries"
+                      indeterminate={
+                        webhooksToDelete.length > 0 && webhooksToDelete.length < webhooks.length
+                      }
+                      value={webhooksToDelete.length === webhooks.length}
+                      onValueChange={value => handleCheckboxChange('main', value)}
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:283]: I would suggest to separate the logic between checking the "main" checkbox and checking the other ones. This will reduce the complexity of the `handleCheckboxChange`. Complexity, in this scenario, corresponds to the number paths existing in the "handleCheckboxChange". 

And for clarity reasons, I would suggest to avoid the conditional in this function since it's arbitrary and note based on the data we have 😊 

Something like:

```js
const handleSelectAllCheckbox = () => {}

const handleSelectOneCheckbox = () => {}
```

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 2]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
+                          <Box paddingLeft={1}>
+                            <IconButton
+                              onClick={() => handleDeleteClick(webhook.id)}
+                              label="Delete"
+                              icon={<DeleteIcon />}
+                              noBorder
+                            />
+                          </Box>
+                        )}
+                      </Row>
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:355]: Instead of doing this, in v9 (or 10) of parts, you can use:

```jsx
<Stack horizontal size={1}>
 <Button1/>
 <Button2/>
</Stack>
```

and it will do the same as what you're trying to do 😊

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 3]
```diff
@@ -276,48 +247,183 @@ function ListView() {
   }
 
   return (
-    <Wrapper>
+    <Layout>
       <PageTitle name="Webhooks" />
-      <Header {...headerProps} />
-      {canRead && (
-        <div className="list-wrapper">
+      <HeaderLayout
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:252]: This component should have an `h1` (`<HeaderLayout as="h1" />`) on it and the page itself should have a `<Main labelledBy="the-id-you-put-on-the-h1">`

Hicham managed this the following way in the registration page 😊 


- Concerning the Main component => https://github.com/strapi/strapi/pull/10733/files#diff-5b8fde11bf3bfe8ba7a3b27f0c5e0d3838892702ff8f1d93ab0589f7351d53abR72
- Concerning the title of the page => https://github.com/strapi/strapi/pull/10733/files#diff-5b8fde11bf3bfe8ba7a3b27f0c5e0d3838892702ff8f1d93ab0589f7351d53abR76

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 4]
```diff
@@ -57,73 +67,13 @@ function ListView() {
...
-        type: 'button',
-      }
-    : null;
-  /* eslint-enable indent */
-
-  const listProps = {
-    title,
-    button: deleteButtonProps,
-    items: webhooks,
+    addBtnLabel: formatMessage({ id: 'Settings.webhooks.list.button.add' }),
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:75]: Could you please pass all the props of this variable directly to the component that is concerned?

Instead of 

```jsx
const headeProps = {
  addBtnLabel: : formatMessage({ id: 'Settings.webhooks.list.button.add' })
}

// and later on

<Button>
{headerProps.addBtnLabel}
</Button>
```

Could you try doing directly:

```jsx
<Button>{formatMessage({ id: 'Settings.webhooks.list.button.add' })}</Button>
```

please 🙏🏻 😊

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 5]
```diff
@@ -267,6 +208,36 @@ function ListView() {
...
+      } else {
+        for (let i = 0; i < webhooksIds.length; i++) {
+          dispatch({
+            type: 'SET_WEBHOOKS_TO_DELETE',
+            value: false,
+            id: webhooksIds[i],
+          });
+        }
+      }
+    }
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:238]: I think this could be simplified:

```js
webhooksIds.forEach(webhook => {
          dispatch({
            type: 'SET_WEBHOOKS_TO_DELETE',
            value: webhooksToDelete.length === 0,
            id: webhook
          })
})
```

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 6]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
-                  />
-                );
-              }}
-            />
+            <Table
+              colCount={5}
+              rowCount={webhooks.length}
+              footer={(
+                <TFooter onClick={() => (canCreate ? handleGoTo('create') : {})} icon={<AddIcon />}>
+                  Add another field to this collection type
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:270]: I think this should go in a translation file 😊 . You have exampelp in this file using the `formatMessage` function

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 7]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
+                </TFooter>
+              )}
+            >
+              <Thead>
+                <Tr>
+                  <Th>
+                    <BaseCheckbox
+                      aria-label="Select all entries"
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:278]: I think you should also put it in a translation file 😊

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 8]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
+                      indeterminate={
+                        webhooksToDelete.length > 0 && webhooksToDelete.length < webhooks.length
+                      }
+                      value={webhooksToDelete.length === webhooks.length}
+                      onValueChange={value => handleCheckboxChange('main', value)}
+                    />
+                  </Th>
+                  <Th>
+                    <TableLabel>name</TableLabel>
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:287]: Same concerning all the `Th` content: I think they should go in a translation file 😛

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 9]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
+                        {webhook.name}
+                      </Text>
+                    </Td>
+                    <Td>
+                      <Text textColor="neutral800">{webhook.url}</Text>
+                    </Td>
+                    <Td>
+                      <Row>
+                        <Switch
+                          onLabel="Enabled"
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:324]: I won't comment about it anymore about it, but every string occurence of a word, in plain english, existing in this file, should go in a translation file 😊

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 10]
```diff
@@ -276,48 +247,183 @@ function ListView() {
...
         onConfirm={handleConfirmDelete}
       />
-    </Wrapper>
+    </Layout>
   );
 }
 
 export default ListView;
+
+// TO KEEP AS EXEMPLE UNTIL WE KNOW HOW TO DEAL WITH EVENTS AND FUNCTIONS DEFINED IN THIS COMPONENT
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:390]: wut? 😆

**Review by mfrachet — CHANGES_REQUESTED** [review]
Overall, this looks great 😄 . I've put comments in the codebase, if something is not clear, feel free to reach out to me 😊 . Great job 👏🏻 🚀

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 11]
```diff
@@ -267,6 +200,36 @@ function ListView() {
     }
   };
 
+  const handleSelectAllCheckbox = () => {
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:203]: In that function, you loop over webhooks 2 times. 
The first time to extract ids. And the second time to select each webhook of the webhookIds.
I think you don't need to do that. You don't need to extract the ids of webhooks. Since you can access the state in the reducer function, you only need to create a function that select all the webhooks.

```js
// Webhooks/ListView/index.js
const handleSelectAllCheckbox = () => {
     dispatch({
      type: 'SET_ALL_WEBHOOKS_TO_DELETE',
    });
};
```

```js
// Webhooks/ListView/reducer.js
...
case 'SET_ALL_WEBHOOKS_TO_DELETE': {
  const { webhooks, webhooksToDelete } = state;

  if (webhooks.length === webhooksToDelete.length) {
    draftState.webhooksToDelete = initialState.webhooksToDelete;
  } else {
    draftState.webhooksToDelete = webhooks.map(webhook => webhook.id);
  }
  break;
}
...
```

Comment (HichamELBSI) [reply packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:203]: Don't forget to add some test to this reducer case 😄

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 12]
```diff
@@ -6,25 +6,34 @@
...
+import { Main } from '@strapi/parts/Main';
+import AddIcon from '@strapi/icons/AddIcon';
+import EditIcon from '@strapi/icons/EditIcon';
+import DeleteIcon from '@strapi/icons/DeleteIcon';
+import EmptyStateDocument from '@strapi/icons/EmptyStateDocument';
 import reducer, { initialState } from './reducer';
+import PageTitle from '../../../components/SettingsPageTitle';
+import adminPermissions from '../../../permissions';
 
 function ListView() {
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:38]: You will need to call the focus hook when the i18n list view will be merged.

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 13]
```diff
@@ -276,47 +239,161 @@ function ListView() {
...
+            </Button>
           )}
-          <ListButton>{canCreate && <Button {...omit(newButtonProps, 'Component')} />}</ListButton>
-        </div>
-      )}
+        />
+
+        <ContentLayout>
+          <>
+            {rowsCount > 0 ? (
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:259]: Something is wrong with the table, I can't focus the delete icon in each row and the last row also (with the keyboard).

webhooksgif

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 14]
```diff
@@ -276,47 +239,161 @@ function ListView() {
...
+              <Table
+                colCount={5}
+                rowCount={webhooks.length}
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:262]: rowCount is already defined on the top of the file : `rowCount={rowsCount}`

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 15]
```diff
@@ -276,47 +239,161 @@ function ListView() {
...
+              >
+                <Thead>
+                  <Tr>
+                    <Th>
+                      <BaseCheckbox
+                        aria-label={formatMessage({
+                          id: 'Settings.webhooks.list.all-entries.select',
+                        })}
+                        indeterminate={
+                          webhooksToDelete.length > 0 && webhooksToDelete.length < webhooks.length
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:280]: Same : `webhooksToDelete.length > 0 && webhooksToDelete.length < rowsCount`

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 16]
```diff
@@ -276,47 +239,161 @@ function ListView() {
...
+                        }
+                        value={webhooksToDelete.length === webhooks.length}
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:282]: Same : `webhooksToDelete.length === rowsCount `

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 17]
```diff
@@ -276,47 +239,161 @@ function ListView() {
...
+                            webhook.name
+                          }`}
+                          value={webhooksToDelete?.includes(webhook.id)}
+                          onValueChange={value => handleSelectOneCheckbox(value, webhook.id)}
+                          id={`${formatMessage({ id: 'Settings.webhooks.list.select' })} ${
+                            webhook.id
+                          }`}
+                          name={`${formatMessage({ id: 'Settings.webhooks.list.select' })} ${
+                            webhook.name
+                          }`}
```
Comment (HichamELBSI) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:321]: The `id` and `name` aren't used to be understood by the user. We should use static `names` and `ids` not translated messages.

**Review by HichamELBSI — CHANGES_REQUESTED** [review]

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js (hunk 13 above)
Comment (ronronscelestes) [reply packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:259]: You can focus on it if you are already focused on the edit icon then using tab. I talked about it with Marvin, if I understood well it's because it's a button/action so you have to use tab to access it 🕵️‍♀️

Comment (HichamELBSI) [reply packages/core/admin/admin/src/pages/Webhooks/ListView/index.js]: Oh thanks for the button, indeed it is better with the tab. But, I can't reach the last list row with the keyboard 😢

Comment (mfrachet) [reply packages/core/admin/admin/src/pages/Webhooks/ListView/index.js]: (rows should include the header)

Comment (ronronscelestes) [reply packages/core/admin/admin/src/pages/Webhooks/ListView/index.js]: Oh right same here, I'm not sure why though! @mfrachet do you have an idea? 🤕

Comment (ronronscelestes) [reply packages/core/admin/admin/src/pages/Webhooks/ListView/index.js]: Fixed 🔥

File: packages/core/admin/admin/src/pages/Webhooks/ListView/index.js [hunk 18]
```diff
@@ -267,56 +202,189 @@ function ListView() {
...
-            />
-          ) : (
-            <EmptyState
-              title={formatMessage({ id: 'Settings.webhooks.list.empty.title' })}
-              description={formatMessage({ id: 'Settings.webhooks.list.empty.description' })}
-              link="https://strapi.io/documentation/developer-docs/latest/development/backend-customization.html#webhooks"
-              linkText={formatMessage({ id: 'Settings.webhooks.list.empty.link' })}
+      <Main labelledBy="webhooks">
+        {isLoading ? (
+          <LoadingIndicatorPage />
```
Comment (mfrachet) [inline packages/core/admin/admin/src/pages/Webhooks/ListView/index.js:228]: Even when displaying the loader, I think the HeaderLayout should be on the page because else, during this tiny moment where nothing is displayed, there's nothing on the screen and the page is in an invalid state because there's not H1 on it.

Also, when loading something, you can put a `aria-busy={isLoading}` on the "Main" component so that it tells people using ATs that something is happening.

Later on, when the data has been loader, I think it's a common practice to use an aria region to announce that the content has been loaded and that the content of the Table is ready to be navigated 😊

**Review by mfrachet — CHANGES_REQUESTED** [review]
I've put a request for changes to investigate the a11y way to handle this page loading :)

**Review by HichamELBSI — APPROVED** [review]
Good job !!
